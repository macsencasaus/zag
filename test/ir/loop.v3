// RUN: %vector -emit-ir %s

fn loop() i64 {
    var x = 0;

    while (x < 1000) {
        ++x;
    }

    return x;
}

fn loop_break() i64 {
    var x = 0;

    while (x < 1000) {
        ++x;
        if (x == 500)
            break;
    }

    return x;
}

fn loop_continue() i64{
    var x = 0;

    while (x < 1000) {
        ++x;
        if (x > 500)
            continue;
        ++x;
    }

    return x;
}
