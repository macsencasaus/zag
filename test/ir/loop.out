; V3 Readable Intermediate Representation
; Function header formatted as:
; @func[<max-stack-size>](<param-type>,...)

@loop[16]():
    %s[0] = i64 0
.L0:
    %s[8] = slt i64 %s[0], i64 1000
    jmpz .L1, i64 %s[8]
    %s[0] = add i64 %s[0], i64 1
    jmp .L0
.L1:
    ret i64 %s[0]

@loop_break[16]():
    %s[0] = i64 0
.L2:
    %s[8] = slt i64 %s[0], i64 1000
    jmpz .L3, i64 %s[8]
    %s[0] = add i64 %s[0], i64 1
    %s[8] = eq i64 %s[0], i64 500
    jmpz .L4, i64 %s[8]
    jmp .L3
.L4:
    jmp .L2
.L3:
    ret i64 %s[0]

@loop_continue[16]():
    %s[0] = i64 0
.L5:
    %s[8] = slt i64 %s[0], i64 1000
    jmpz .L6, i64 %s[8]
    %s[0] = add i64 %s[0], i64 1
    %s[8] = sgt i64 %s[0], i64 500
    jmpz .L7, i64 %s[8]
    jmp .L5
.L7:
    %s[0] = add i64 %s[0], i64 1
    jmp .L5
.L6:
    ret i64 %s[0]

